feat: support tailwindcss v4

winchesHe opened the pull request:
<!---
Thanks for creating a Pull Request ❤️!

Please read the following before submitting:
- PRs that adds new external dependencies might take a while to review.
- Keep your PR as small as possible.
- Limit your PR to one type (docs, feature, refactoring, ci, repo, or bugfix)
-->

Closes #4644

## Description

The current migration is working fine with `tailwindcss js config`.

It won't be released that soon because it has a lot of breaking changes.

### Need to update

1. `tailwind-merge` upgrade ✅
2. ~codemod support `heroui` plugin transform to `css` file~

### Migration steps

1. Follow the official upgrade guide `https://tailwindcss.com/docs/upgrade-guide`
2. Install the preview version https://github.com/heroui-inc/heroui/pull/4656#issuecomment-2651218074
3. Use heroui with `@config "path/to/tailwindcss"`

## Heroui template for tailwindcss v4

https://github.com/heroui-inc/vite-template/pull/10

jprosevear commented on apps/docs/app/examples/perf/client-page.tsx:
```diff
@@ -101,7 +101,7 @@ const MyInput = extendVariants(Input, {
           "bg-zinc-100",
           "border",
-          "shadow",
+          "shadow-xs",
```
I think shadow changed to shadow-sm https://tailwindcss.com/docs/upgrade-guide#renamed-utilities

winchesHe replied:
yeah agree 👍

jprosevear commented on apps/docs/content/landing/index.tsx:
```diff
@@ -61,7 +61,7 @@ export default {
       title: "Focus interactions",
       description:
-        "Focus ring will appear only when user navigates with keyboard or screen reader.",
+        "Focus ring-3 will appear only when user navigates with keyboard or screen reader.",
```
Search and replace error?

winchesHe replied:
nice catch thank you guy! 👍

jprosevear commented:
@winchesHe nice; I learned some stuff :-)

sapjax commented:
Can you release a pre version of this PR? 
Then the users who have migrated to tailwind v4 can use it first, and it doesn't need to be perfect, just need to work in most cases.

TianMengLucky commented:
What's the progress on this PR?

winchesHe commented:
> Can you release a pre version of this PR?你能发布这个 PR 的预版本吗？ Then the users who have migrated to tailwind v4 can use it first, and it doesn't need to be perfect, just need to work in most cases.然后迁移到 tailwind v4 的用户可以先使用它，而且不需要完美，只需要在大多数情况下工作即可。

pull request preview version will be released once https://github.com/heroui-inc/heroui/pull/4540 is up and running

jrgarciadev commented:
Fix the conflicts @winchesHe

Butonix commented:
I installed the preview version, but the prefix is ​​not loading

**let's say**
```
.rounded-large {
    border-radius: var(--heroui-radius-large);
}
```

Property --heroui-radius-large is not defined

Butonix commented:
There is no cursor-pointer on the buttons

Franac1 commented:
https://github.com/heroui-inc/heroui/pull/4656#issuecomment-2651218074

Hej, how exactly do i install all those dependencies? I have problem with resolving dependencies like:
npm warn Could not resolve dependency:
npm warn peer @heroui/theme@">=2.4.7" from @heroui/calendar@2.2.10

brandonhenness commented:
> [#4656 (comment)](https://github.com/heroui-inc/heroui/pull/4656#issuecomment-2651218074)
> 
> Hej, how exactly do i install all those dependencies? I have problem with resolving dependencies like: npm warn Could not resolve dependency: npm warn peer @heroui/theme@">=2.4.7" from @heroui/calendar@2.2.10

I'm having the same issue. I was able to get it to work by using the `--legacy-peer-deps` flag but that's not a clean solution as an end user can't install using `npm install` without usign the flag. Currently, I'm trying to install dirrect from the [feat/tailwindcss-v4](https://github.com/heroui-inc/heroui/tree/feat/tailwindcss-v4) branch but's taking **FOREVER**. If I get it to work I'll come back and let you know.

The command is supposed to be `npm i "github:heroui-inc/heroui#feat/feat/tailwindcss-v4"` if you want to give it a try yourself. If this doesn't work then maybe hero ui could merge these changes into the beta branch and then create a beta dist-tag on npm?

brandonhenness commented:
> Currently, I'm trying to install directly from the [feat/tailwindcss-v4](https://github.com/heroui-inc/heroui/tree/feat/tailwindcss-v4) branch, but it's taking **FOREVER**. If I manage to get it working, I'll come back and let you know.

I was not able to get this to work either.

> If this doesn't work, perhaps Heroui could merge these changes into the beta branch and create a beta dist-tag on npm?

@winchesHe what are the chances of getting a beta dist-tag on npm once this is merged with the [beta/release-next](https://github.com/heroui-inc/heroui/tree/beta/release-next) branch?

Franac1 commented:
> > Currently, I'm trying to install directly from the [feat/tailwindcss-v4](https://github.com/heroui-inc/heroui/tree/feat/tailwindcss-v4) branch, but it's taking **FOREVER**. If I manage to get it working, I'll come back and let you know.
> 
> I was not able to get this to work either.
> 
> > If this doesn't work, perhaps Heroui could merge these changes into the beta branch and create a beta dist-tag on npm?
> 
> @winchesHe what are the chances of getting a beta dist-tag on npm once this is merged with the [beta/release-next](https://github.com/heroui-inc/heroui/tree/beta/release-next) branch?

Thank you for saving me time. 
I'm trying to use their killer date range picker component. I can't find anywhere else as near as good as theirs.
I really hope they start supporting tw v4.

winchesHe commented:
> @winchesHe what are the chances of getting a beta dist-tag on npm once this is merged with the [beta/release-next](https://github.com/heroui-inc/heroui/tree/beta/release-next) branch?@winchesHe 一旦它与 [beta/release-next](https://github.com/heroui-inc/heroui/tree/beta/release-next) 分支合并，在 npm 上获得 beta dist-tag 的机会有多大？

Once merged will release a latest beta version in npm

winchesHe commented:
https://github.com/tailwindlabs/tailwindcss/issues/16664 related `ease-out` important modifier issue

christophmeise commented:
When I use any of these preview versions I get an runtime error:

```
Parsing css source code failed
  2979 |   }
  2980 |   .w-\[calc\(100\%_-_var\(spacing\.6\)\)\] {
> 2981 |     width: calc(100% - var(spacing.6));
       |                             ^
  2982 |   }
  2983 |   .w-\[calc\(2623\/16\*1rem\)\] {
  2984 |     width: calc(2623 / 16 * 1rem);

Unexpected token Ident("spacing") at [project]/apps/nextjs/src/styles/globals.css:2980:28
```

I can confirm that this error goes away as soon as I remove the heroui content import from the tailwind config: ../../node_modules/@heroui/theme/dist/**/*.{js,ts,jsx,tsx}
I also removed all CSS and plugins to confirm that this is a problem with heroui.

Using Tailwind 4.0.9, Nextjs 15.0.2 and tried both of the latest beta preview versions. I checked the implementation of the vite example and I am doing it exactly like that.

winchesHe commented:
> When I use any of these preview versions I get an runtime error:
> 
> ```
> Parsing css source code failed
>   2979 |   }
>   2980 |   .w-\[calc\(100\%_-_var\(spacing\.6\)\)\] {
> > 2981 |     width: calc(100% - var(spacing.6));
>        |                             ^
>   2982 |   }
>   2983 |   .w-\[calc\(2623\/16\*1rem\)\] {
>   2984 |     width: calc(2623 / 16 * 1rem);
> 
> Unexpected token Ident("spacing") at [project]/apps/nextjs/src/styles/globals.css:2980:28
> ```
> 
> I can confirm that this error goes away as soon as I remove the heroui content import from the tailwind config: ../../node_modules/@heroui/theme/dist/**/*.{js,ts,jsx,tsx} I also removed all CSS and plugins to confirm that this is a problem with heroui.
> 
> Using Tailwind 4.0.9, Nextjs 15.0.2 and tried both of the latest beta preview versions. I checked the implementation of the vite example and I am doing it exactly like that.

Fixed in the latest version https://github.com/heroui-inc/heroui/runs/37994128721

aeltorio commented:
@winchesHe ,
Thank you for your impressive work.  
I've updated the template React 19, TailwindCSS 4.0.9, ESLint 9, Heroui 2.7… with 
https://github.com/sctg-development/vite-react-heroui-template
Thanks a lot

Ankan002 commented:
Does this also support nextjs?

aeltorio commented:
I have synced my [template](https://github.com/sctg-development/vite-react-heroui-template) to @winchesHe c305b94 build.  
For our needs, I have added i18n support and a simple language-switching component to change the language. This component correctly handles both LTR and RTL languages. Please note that the languages are automatically translated.

<img width="1161" alt="image" src="https://github.com/user-attachments/assets/de3dd8a9-6c1d-4a84-9a34-84430c25e17e" />
<img width="1162" alt="image" src="https://github.com/user-attachments/assets/975fb30c-d54b-495b-a563-ced41e6c1ad1" />

The template without i18n ends at [e8d9065d36ef09c90714b099406ee6a737ef32c3](https://github.com/sctg-development/vite-react-heroui-template/tree/e8d9065d36ef09c90714b099406ee6a737ef32c3)

sapjax commented:
I found that heroui doesn't support [using-a-data-attribute](https://tailwindcss.com/docs/dark-mode#using-a-data-attribute) as dark mode selector
```
@custom-variant dark (&:where([data-theme=dark], [data-theme=dark] *));
```

It only works with class selector:
```
@custom-variant dark (&:where(.dark, .dark *));
```
